synedrion/src/www02/bip32.rs

```diff
@@ -0,0 +1,206 @@
+    /// Return a verifying key derived from the given type using the BIP-32 scheme.
+    fn derive_verifying_key_bip32(&self, derivation_path: &DerivationPath) -> Result<VerifyingKey<C>, bip32::Error>;
+}
+
+mod sealed {
+    use super::*;
+    pub trait Sealed {}
+    impl Sealed for VerifyingKey<tiny_curve::TinyCurve64> {}
+    impl Sealed for VerifyingKey<k256::Secp256k1> {}
+    impl Sealed for SigningKey<tiny_curve::TinyCurve64> {}
+    impl Sealed for SigningKey<k256::Secp256k1> {}
+}
+
+/// Trait for types that can derive BIP32 style "tweaks" from public keys.
+pub trait PubTweakable: sealed::Sealed {
+    fn tweakable_pk(&self) -> impl bip32::PublicKey + Clone;
+}
+
+/// Trait for types that can derive BIP32 style "tweaks" from secret keys.
+pub trait SecretTweakable: sealed::Sealed {
+    /// Convert `self` into something that can be used for BIP32 derivation.
+    fn tweakable_sk(&self) -> impl bip32::PrivateKey + Clone;
+}
+
+impl PubTweakable for VerifyingKey<<TestParams as SchemeParams>::Curve> {
+    fn tweakable_pk(&self) -> impl bip32::PublicKey + Clone {
+        let pk: PublicKey<_> = self.into();
+        let wrapped_pk: PublicKeyBip32<_> = pk.into();
+        wrapped_pk
+    }
+}
+
+impl PubTweakable for VerifyingKey<<ProductionParams112 as SchemeParams>::Curve> {
+    fn tweakable_pk(&self) -> impl bip32::PublicKey + Clone {
+        *self
+    }
+}
+
+impl SecretTweakable for SigningKey<<TestParams as SchemeParams>::Curve> {
+    fn tweakable_sk(&self) -> impl bip32::PrivateKey + Clone {
+        let sk: SecretKey<_> = self.into();
+        let wrapped_sk: PrivateKeyBip32<_> = sk.into();
+        wrapped_sk
+    }
+}
+
+impl SecretTweakable for SigningKey<<ProductionParams112 as SchemeParams>::Curve> {
+    fn tweakable_sk(&self) -> impl bip32::PrivateKey + Clone {
+        let sk: SecretKey<_> = self.into();
+        sk
+    }
+}
+
+impl<P, I> ThresholdKeyShare<P, I>
+where
+    P: SchemeParams,
+    VerifyingKey<P::Curve>: PubTweakable,
+    SigningKey<P::Curve>: SecretTweakable,
+    I: Clone + Ord + PartialEq + Debug + Serialize + for<'x> Deserialize<'x>,
+{
+    /// Deterministically derives a child share using BIP-32 standard.
+    pub fn derive_bip32(&self, derivation_path: &DerivationPath) -> Result<Self, bip32::Error> {
+        let pk = self.verifying_key().map_err(|_| bip32::Error::Crypto)?;
+        let tweakable_pk = pk.tweakable_pk();
+        let tweaks = derive_tweaks::<P::Curve>(&tweakable_pk, derivation_path)?;
+
+        // Will fail here if secret share is zero
+        let secret_share = self.secret_share.clone().to_signing_key().ok_or(bip32::Error::Crypto)?;
+        let mut tweakable_sk = secret_share.tweakable_sk();
+        for tweak in &tweaks {
+            tweakable_sk = tweakable_sk.derive_child(*tweak)?;
+        }
+        let bytes = tweakable_sk.to_bytes();
+        let bytes = bytes.get(32 - Scalar::<P>::repr_len()..).ok_or(bip32::Error::Decode)?;
```

Comment:
Wouldn't it be easier to just unwrap `SecretTweakable` into `SecretKey`, via another method in the `SecretTweakable` trait?

Reply:
Can you elaborate a bit on this (and "I don't think you need to de/serialize at all" above), not sure how you mean!

Reply:
See https://github.com/fjarri/synedrion/commit/fa153228a3231c81a66baefaace6e44fe713f7a1 (naming is rough, bounds may be adjusted, and perhaps some arguments can be taken by value). Basically, you have a trait to convert `VerifyingKey<C>` to something `bip32` accepts, so I just add another method to convert that something back to `VerifyingKey<C>`, which we can do either by just returning the same object (for `k256`) or unwrapping the newtype (for `tiny-curve`). Same for `SigningKey`.

Reply:
Cherry picked! ty